File: cpp/shipment.cpp
#include "shipment.h"



shipment::shipment()
{

}

shipment::shipment(customer cust,Transport transport,std::vector<Product> productsPackaged){
    this->cust = cust;
    this->transport = transport;
    this->productsPackaged = productsPackaged;
    this->dispatchDate = std::chrono::system_clock::now() + std::chrono::hours(24 * 2);
    this->deliveryDate = std::chrono::system_clock::now() + std::chrono::hours(24 * 2 * (int)calculateDeliveryTime());
    this->status = "Dispatching";

}
shipment::~shipment(void){

}

customer shipment::getCustomer() const{
    return cust;
}

Transport shipment::getTransport() const{
    return transport;
}

double shipment::getTotalUnits() const{
    return totalUnits;
}

void shipment::setCustomer(customer cust){
    this->cust = cust;
}

void shipment::setTransport(Transport transport){
    this->transport = transport;
}

void shipment::setTotalUnits(double units){
    this->totalUnits = units;
}

void shipment::addShipment(const shipment& shipments, std::vector<shipment> &array){
    array.push_back(shipments);
}

void shipment::deleteShipment(std::vector<shipment> &array,int pos){
    array.erase(array.begin()+pos);
}

void shipment::displayInfo(){

    cout <<  "Customer: " << cust.getName() << ", Transport : " << transport.getName() << "\n";
    cout << "List of Products: \n";
    for(Product products : this->productsPackaged){
        products.displayInfo();
        cout << "\n";
    }
    cout << "Total price of the products: " << calculateProductsPrice() << "\n";
    cout << "Total price of the transport: " << calculateTransportCost() << "\n";
    cout << "Total price of the packagings: " << getTotalUnits() << "\n";

    std::time_t dispatchDate_t = std::chrono::system_clock::to_time_t(dispatchDate);
    std::tm* printedDispatchDate= std::localtime(&dispatchDate_t);

    std::time_t deliveryDate_t = std::chrono::system_clock::to_time_t(deliveryDate);
    std::tm* printedDeliveryDate= std::localtime(&deliveryDate_t);

    cout << "Expected Date of Distatch: " << std::put_time(printedDispatchDate, "%Y-%m-%d %H:%M:%S");
    cout << "  -  Expected Time of Delivery: " << std::put_time(printedDeliveryDate, "%Y-%m-%d %H:%M:%S");

    cout << "\n\nStatus: " << status << "\n";
}

void shipment::addProduct(const Product product){
    bool added = false;
    if(this->productsPackaged.empty()){
        this->productsPackaged.push_back(product);
        added = true;
    }else{
        for(int i=0;i<(int)this->productsPackaged.size();i++){
            if(this->productsPackaged[i].equals(product)){
                this->productsPackaged[i].setPackagingUnits(this->productsPackaged[i].getPackagingUnits()+product.getPackagingUnits());
                added = true;
                break;
            }
        }
        if(!added){
            this->productsPackaged.push_back(product);
        }
    }
}

bool shipment::isDispatched(){

    auto currentTime = std::chrono::system_clock::now();


    return (currentTime > this->dispatchDate);
}

void shipment::cancel(){


}

double shipment::calculateProductsPrice(){
    double total =0;
    for(Product products : this->productsPackaged){
        total +=products.getPrice()*products.getPackagingUnits();
    }

    return total;

}

double shipment::calculateDeliveryTime(){
    return this->cust.getDistance() / this->transport.getSpeed();
}

double shipment::calculateTransportCost(){
    return this->cust.getDistance() * this->transport.getCost();
}

int shipment::calculateProductsQuantity(){
    int total = 0;

    for(Product products : this->productsPackaged){
        total +=products.getPackagingUnits();
    }

    return total;
}

File: java/shipment.java
package cps2004.java;

import java.util.List;
import java.util.Date;
import java.text.SimpleDateFormat;

public class shipment {

    private customer cust;
    private transport transport;
    private List<products> productsPackaged;
    private Date dispatchDate;
    private Date deliveryDate;
    private String status;
    private double totalUnits;
    private double totalCost;

    // Constructors
    public shipment() {}

    public shipment(customer cust, transport transport, List<products> productsPackaged) {
        this.cust = cust;
        this.transport = transport;
        this.productsPackaged = productsPackaged;
        this.dispatchDate = new Date(System.currentTimeMillis() + (24 * 2 * 60 * 60 * 1000)); // 2 days
        this.deliveryDate = new Date(System.currentTimeMillis() + (24 * 2 * 60 * 60 * 1000) +((int)calculateDeliveryTime()*60 * 60 * 1000));
        this.status = "Dispatching";
    }

    // Getters & Setters
    public customer getCustomer() {
        return cust;
    }

    public transport getTransport() {
        return transport;
    }

    public List<products> getProductsPackaged(){
        return productsPackaged;
    }

    public double getTotalUnits(){
        return totalUnits;
    }
    
    public double getTotalPrice() {
        return totalCost;
    }
    public void setCustomer(customer cust) {
        this.cust = cust;
    }

    public void setTransport(transport transport) {
        this.transport = transport;
    }

    public void setTotalUnits(double units){
        this.totalUnits = units;
    }

    public void setStatus(String status){
        this.status = status;
    }

    public void setTotalPrice(double total){
        this.totalCost = total;
    }

    // Other Methods
    public void addShipment(shipment shipment, List<shipment> array) {
        array.add(shipment);
    }

    public void deleteShipment(List<shipment> array, int pos) {
        array.remove(pos);
    }

    public void displayInfo() {
        System.out.println("Customer: " + cust.getName() + ", Transport: " + transport.getName());
        System.out.println("List of Products: ");
        for (products product : this.productsPackaged) {
            product.displayInfo();
            System.out.println();
        }
        System.out.println("Total price of the products: " + calculateProductsPrice());
        System.out.println("Total price of the transport: " + calculateTransportCost());
        System.out.println("Total price of the packagings: " + getTotalUnits());

        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        System.out.println("Expected Date of Dispatch: " + sdf.format(dispatchDate) +
                " - Expected Time of Delivery: " + sdf.format(deliveryDate));

        System.out.println("\nStatus: " + status);
    }

    public void addProduct(products product) {
        boolean added = false;
        if (this.productsPackaged.isEmpty()) {
            this.productsPackaged.add(product);
            added = true;
        } else {
            for (int i = 0; i < this.productsPackaged.size(); i++) {
                if (this.productsPackaged.get(i).equals(product)) {
                    this.productsPackaged.get(i).setPackagingUnits(
                            this.productsPackaged.get(i).getPackagingUnits() + product.getPackagingUnits());
                    added = true;
                    break;
                }
            }
            if (!added) {
                this.productsPackaged.add(product);
            }
        }
    }

    public boolean isDispatched() {
        return System.currentTimeMillis() > this.dispatchDate.getTime();
    }

    public void cancel() {
        // Implementation for cancellation
    }

    public double calculateProductsPrice() {
        double total = 0;
        for (products product : this.productsPackaged) {
            total += product.getPrice() * product.getPackagingUnits();
        }
        return total;
    }

    public double calculateDeliveryTime() {
        return this.cust.getDistance() / this.transport.getSpeed();
    }

    public double calculateTransportCost() {
        return this.cust.getDistance() * this.transport.getCost();
    }

    public void updateShipments(List<products> auxList, customer customer, transport transport,double total) {
        for(int i=0;i<auxList.size();i++){
            this.addProduct(auxList.get(i));
        }
        this.setCustomer(customer);
        this.setTransport(transport);
        this.setTotalPrice(total);
    }

}
